### ISSUE TITLE
hbw_malloc not working with default MPOL_PREFERRED mode on Centos 7.4 and older

### ISSUE THREAD

**Issue by svalat** [memkind/memkind#47]
Hi,
after searching I just found that Centos 7.4 and older has a kernel bug which didn't follow the MPOL_PREFERRED constrain when using huge pages (which is default now):
 * https://lkml.org/lkml/2015/6/18/690
 * https://access.redhat.com/solutions/3155591

So it make the hbw_malloc ineficient except if forcing policy to BIND mode : 

```
hbw_set_policy(HBW_POLICY_BIND)
```

This would be nice to notice it into the memkind documentation as a user might think he made an MCDRAM allocation on KNL but will not in practice (I made the test on KNL in Flat + SNC4 mode and on centos 7.4.1708).

It can be shown with : 

```c
#include <hbwmalloc.h>
#include <stdio.h>
#include <numaif.h>
#include <sys/mman.h>
#include <string.h>

//Compile with : gcc main.c -O0 -I/MEMKIND/include/ -L/MEMKIND/lib/ -lmemkind -lnuma
//Run :
//./a.out [bind]

//assign to NUMA node 4, MCDRAM (considering KNL in FLAT + SNC4), change to 0x4 if you are in SNC2 mode and 0x1 in non SNC.
#define MCDRAM_BIND 0x10
//#define MCDRAM_BIND 0x4
//#define MCDRAM_BIND 0x1
#define SIZE (20*1024*1024)

void run(char * data, const char * descr)
{
	int i;

	//first touch
	for (i = 0 ; i < SIZE ; i++)
		data[i] += 1;

	//info
	printf("\n%s :\n");

	size_t cnt[8];
	memset(cnt,0,sizeof(cnt));

	//check page mapping
	for (i = 0 ; i < SIZE ; i+=4096)
	{
		int status;
		void * pages[1] = {(void*)((size_t)(data+i) & ~(4095))};
		int res = move_pages(0,1,pages,NULL,&status,0);
		if (res == 0)
			cnt[status]++;
	}

	//display
	for (i = 0 ; i < 8 ; i++)
			printf("Pages on %d : %lu\n",i,cnt[i]);
}

int main(int argc, char ** argv)
{
	printf("status = %d\n",hbw_check_available());

	//if not set hbw_malloc seems to not put pages on MCDRAM
	if (argc > 1 && strcmp(argv[1],"bind") == 0)
	{
		hbw_set_policy(HBW_POLICY_BIND);
		printf("Seth hbw_policy_bind\n");
	}

	char * data = malloc(SIZE);
	run(data, "std malloc");

	data = hbw_malloc(SIZE);
	run(data, "hbw_malloc");

	unsigned long a = MCDRAM_BIND;
	data = mmap(NULL,SIZE,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANONYMOUS,0,0);
	mbind((void*)((size_t)(data) & ~(4095)), SIZE, MPOL_BIND,&a, sizeof(unsigned long)*8,0);
	run(data,"mbind with BIND");

	data = mmap(NULL,SIZE,PROT_READ|PROT_WRITE,MAP_PRIVATE|MAP_ANONYMOUS,0,0);
	mbind((void*)((size_t)(data) & ~(4095)), SIZE, MPOL_PREFERRED,&a, sizeof(unsigned long)*8,0);
	run(data,"mbind with PREFERRED");
}
```

Output : 
```
status = 0

std malloc :
Pages on 0 : 5120
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 0
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0


hbw_malloc :
Pages on 0 : 5120  <-- should be on node 4
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 0 <-- should be here
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0

mbind with BIND :
Pages on 0 : 0
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 5120
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0

mbind with PREFERRED :
Pages on 0 : 5120  <-- should be on node 4
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 0 <-- should be here
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0
```

**Comment by ArturKoziej** [comment]
Hi Sébastien,
Thanks for note. We will consider to document it.

You can try to disable THP (https://www.kernel.org/doc/Documentation/vm/transhuge.txt). This should restore the original behavior of MPOL_PREFERRED, so the hbwmalloc policy change would not be necessary.

Cheers,
Artur

**Comment by svalat** [comment]
I will check if I can ask our admin to make the test and confirm. I'll keep you in touch.
Notice that there should be a fix on kernel 4.X (https://lkml.org/lkml/2015/6/18/690) don't know which one. So it is more a temporary bug.

**Comment by svalat** [comment]
Ok, I made the test by disabling huge pages on our KNL and it fixed the issue.

Output without huge pages :

```
status = 0

std malloc :
Pages on 0 : 5120
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 0
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0

hbw_malloc :
Pages on 0 : 0
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 5120 <-- That's ok now
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0

mbind with BIND :
Pages on 0 : 0
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 5120
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0

mbind with PREFERRED :
Pages on 0 : 0
Pages on 1 : 0
Pages on 2 : 0
Pages on 3 : 0
Pages on 4 : 5120 <-- That's ok now
Pages on 5 : 0
Pages on 6 : 0
Pages on 7 : 0
```

**Comment by ArturKoziej** [comment]
Nice to hear, thanks!
Red Hat seem to know the issue and hopefully they will fix it soon.

**Comment by grzegorz-andrejczuk** [comment]
The issue should be fixed in kernel 3.10.0-693.2.1. 
It is available in CentOS repos:
http://mirror.centos.org/centos/7.4.1708/updates/x86_64/Packages/kernel-3.10.0-693.2.1.el7.x86_64.rpm 

Best Regards,
Grzegorz

**Comment by lukaszstolarczuk** [comment]
This issue is a bit stale for a while, so I hope it is no more a problem. If you feel like continuing, please just re-open it or file a new one.
